src/jtag/drivers/bcm2835gpio.c

```diff
@@ -93,10 +93,12 @@ static int speed_offset = 28;
 static unsigned int jtag_delay;
 
 static const struct adapter_gpio_config *adapter_gpio_config;
-static struct initial_gpio_state {
-	unsigned int mode;
-	unsigned int output_level;
-} initial_gpio_state[ADAPTER_GPIO_IDX_NUM];
+static struct gpio_pin_config {
+	unsigned int initial_mode;
+	unsigned int initial_output_level;
+	unsigned int register_offset;
+	unsigned int shift_amount;
+} gpio_pin_config[ADAPTER_GPIO_IDX_NUM];
```

Comment:
I was initially onboard with this idea, because, like you, I don't like meaningless, repetitive math, and it seems to make sense to calculate this value once and use it over and over again...

But, I don't know that we're actually reducing any overhead by doing this. With standard -O2 compiler optimizations, the divide gets assembled to:
```
        lsr     w2, w3, 5
``
and the modulo to:
```
        lsl     w1, w1, w3
```

these are just shifts and not complex divides (thanks to the divisor being 32) so should be rather quick in practice.

However, we're now replacing these with memory reads to look up the shift and offset values from the array of structs which may be more expensive than a 1-2 cycle arithmetic shift.

Anyway, all that is to say... without some measurable/obvious performance gain from this change, I think we should maybe not do this.

If you have benchmarks showing a statistically significant performance benefit from this change, then maybe it's ok to send upstream.